DataViews: Enhance filter operator handling to support field-type-specific labels

chihsuan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

## Proposed Changes

Related to https://github.com/Automattic/wp-calypso/pull/103525#discussion_r2096989610

This PR refactors the operator handling logic in the DataViews package to allow field types to define user-friendly, context-specific labels for filter operators.

To enable this, the `FilterByConfig` is updated to accept an array of either operator keys or OperatorConfig objects, allowing for flexible and descriptive operator labels.


```ts
export interface OperatorConfig {
	name: Operator;
	label: string;
}

export interface FilterByConfig {
	/**
	 * The list of operators supported by the field.
	 * Can be an array of operator keys, or an array of objects mapping operator keys to their config (e.g. label).
	 */
	operators?: ( OperatorConfig | Operator )[];

	/**
	 * Whether it is a primary filter.
	 *
	 * A primary filter is always visible and is not listed in the "Add filter" component,
	 * except for the list layout where it behaves like a secondary filter.
	 */
	isPrimary?: boolean;
}
```

Refactor the operator normalization logic to ensure all operators are consistently structured.

```ts
export interface NormalizedFilter {
...
	/**
	 * The list of operators supported by the field.
	 */
	operators: OperatorConfig[];
}
```

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

The previous implementation limited filter operator labels to static, generic values, which could be less user-friendly for certain field types (e.g., dates).

For example, date fields can now use labels like "before", "after", or "on or after" instead of generic "less than" or "greater than". Or number/currency fields can use labels like "equal to", "not equal to" instead of "is" or "is not".

The refactor also improves maintainability and makes it easier to add or modify operators in the future.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run yarn dataviews:storybook:start
- Go to http://localhost:57863/?path=/story/dataviews-dataviews--default
- Select the date filter
- Selec the "before (inc)" operator and verify that the filter is applied correctly
- Try other operators, e.g. `after (inc)`


https://github.com/user-attachments/assets/d220f036-1675-4725-a6bc-1796b24fa5ee

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

chihsuan committed:
Refactor filter operator handling to use structured operator configuration

- Introduced `OperatorConfig` interface to encapsulate operator name and label.
- Updated `FilterByConfig` to support an array of `OperatorConfig` or operator keys.
- Modified `sanitizeOperators` function to normalize operators based on the new structure.
- Adjusted components to utilize the new operator configuration, ensuring consistent operator label usage across filters.
- Cleaned up unused operator checks in `FilterText` component for improved readability.

chihsuan committed:
Fix type

chihsuan committed:
Fix type

chihsuan committed:
Add changelog

youknowriad commented:
I'm not sure I like this being a configurable thing,

In the examples you give (date, currency...) ... I think DataViews internally should render the right labels, either by making these different operators entirely or by having a way to have multiple labels but I don't think the user of DataViews should have to think about giving labels to operators.

chihsuan committed:
Update changelog

chihsuan committed:
Merge branch 'trunk' into enhance/custom-filter-operator-label

chihsuan commented:
I've changed the PR title and description to better reflect the goal of this enhancement.

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -24,11 +24,17 @@ export interface Option< Value extends any = any > {
+	 * Can be an array of operator keys, or an array of objects mapping operator keys to their config (e.g. label).
 	 */
-	operators?: Operator[];
+	operators?: ( OperatorConfig | Operator )[];
```
Should this stay `Operator[]` then?

youknowriad replied:
I guess I'm not understanding what this FilterByConfig type applies to? is it the "Field type" or the "Field Type type"

youknowriad replied:
I see that `FilterByConfig` is used by both, I think this is where we should separate to two distinct types maybe.

chihsuan replied:
Great point! Yeah, this way, we can customize labels within the field type definition while preventing users from modifying them directly.

chihsuan committed:
Add 'between' operator and enhance filterBy configuration

- Introduced OPERATOR_BETWEEN constant in constants.ts.
- Updated normalizeFields function to support custom operator labels.
- Modified FilterByConfig to allow for operator label overrides.
- Adjusted datetime field type to include operator label overrides for less than and greater than or equal operators.

chihsuan committed:
Revert unrelated change

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -203,6 +216,9 @@ export type NormalizedField< Item > = Omit< Field< Item >, 'Edit' > & {
 	isValid: ( item: Item, context?: ValidationContext ) => boolean;
 	enableHiding: boolean;
 	enableSorting: boolean;
+	filterBy: Omit< FilterByConfig, 'operators' > & {
```
Should we simplify this and just normalize this to `OperatorConfig[]`

chihsuan replied:
Sounds good! Updated in 05f3718b8af138cf391773c25f13d6c67ddf26f4

chihsuan committed:
Changed FilterByConfig to require operators as an array of OperatorConfig

oandregal commented on packages/dataviews/src/field-types/datetime.tsx:
```diff
@@ -42,5 +58,9 @@ export default {
 	enableSorting: true,
 	filterBy: {
 		operators,
+		overrideOperatorLabels: {
```
Instead of creating a separate key to override the label, what if we make `filterBy.operators` take either a string (key) or the operator definition (key, label)?

oandregal replied:
By doing this, the consumer would also be able to override them via the field declaration as well.

chihsuan replied:
That was the initial implementation, and I think it works well. However, we decided to go with the current implementation because we don't think customers would need to override the operators.

https://github.com/Automattic/wp-calypso/pull/103660#issuecomment-2904682558

oandregal commented:
I just came from AFK and wasn't able to review the work for operators that landed last week, bear with me while I catch up 🙇 

> For example, date fields can now use labels like "before", "after", or "on or after" instead of generic "less than" or "greater than". Or number/currency fields can use labels like "equal to", "not equal to" instead of "is" or "is not".

My primary question is if an operator such as "less than" can work with both integer and datetime types. It connects with [this related thought](https://github.com/Automattic/wp-calypso/pull/103577#discussion_r2107632471) about operators and the filtering implementation , and also the relationship between [operators and filter/Edit UI](https://github.com/Automattic/wp-calypso/pull/103577#discussion_r2107647169).

Without having looked at the code yet, I think the question we should ask is how this is better than having operators that are type-aware: having both the "less than" operator (integer) and the "before" operator (date), etc.

chihsuan committed:
Merge branch 'trunk' into enhance/custom-filter-operator-label

chihsuan committed:
Merge branch 'trunk' into enhance/custom-filter-operator-label

oandregal commented:
I've got to catch up with the current state of operators. I think we should do the following:

- Curate which operators are valid per field type. At the moment, you can get weird mixes: what does it mean to have a "boolean field type with a great than operator"? I've prepared [#104040](https://github.com/Automattic/wp-calypso/pull/104040) to address this.
- The examples I've seen that this PR would serve (before/after, equal) can be modeled by introducing new operators instead. I favor that approach for now.

oandregal commented:
https://github.com/Automattic/wp-calypso/pull/104040 has been deployed.

chihsuan commented:
Closing this. I created another PR instead. https://github.com/Automattic/wp-calypso/pull/104094
